one year post

ollieparanoid opened the pull request:
* [Preview screenshot](https://user-images.githubusercontent.com/6033024/41129071-06bec508-6aa0-11e8-9a98-b25803b39272.png) (updated 2018-06-07)
* [Preview rendered by GitHub](https://github.com/postmarketOS/postmarketos.org/blob/one-year/content/blog/2018-06-09-one-year.md) (destroyed layout but clickable links)

This is a first draft, lots of content is missing. The intro is quite different from the other blog posts to keep it more interesting. Mainline is a complex topic, so I wouldn't be surprised if this contains some errors or credits to some people are missing. Please let me know about that and how you like it. Looking forward for feedback!

(As written in the other issue, if we can't make the exact date, I'm for taking the extra days to make it a nice post instead of putting something out there with less quality.)

## Help wanted!
For some media, we already have nice photos that only need to be found in either the wiki or in Riot and linked here. That would already be helpful!

### Open Questions
* @opendata26: as I understand it, modem should work in theory with the sirius as well, but it has not been tried out yet, right?

### Media
* More Nexus 5 photos to replace the duplicated ones in the current version:
  * One showing the kernel version (@MartijnBraam did that already)
  * How about one with nheko with its dark theme and qt5-qtvirtualkeyboard?
  * Doing more with the modem in plasma mobile if that is possible
* Photos of the sirius and klte (@opendata26 and @drebrez, not sure if it makes sense with klte since display isn't working)
* Maybe a few more device photos with the new wallpaper. The raspberry pi one is awesome, I'd like to include that for sure.
* Create a short video / gif of the "maximum-attention" initramfs hook (see the other blog posts for reference) (@MayeulC)
* i3wm photos on the N900 (one with the fixed battery indicator would be nice)
* matchbox and related PDA applications photos (@duncanguthrie)
* A graph for packaged devices over time:
  * writing a script that outputs the count of `aports/device/device-*` folders for each day that the project has been running
  * finding a way to convert that to a graph (libreoffice, graphviz script, ...)

### Content
* If someone could create the list of new devices (see previous posts for reference), that would be a huge help (maybe @zhuowei has time for that again?)
* @bhush9: you wanted to write a few words about the plasma mobile status - is that still planned, and what would you write about?
* Is there content I forgot? Please let me know as well!

Closes #56.

MayeulC commented:
All right, that was an awesome ride (even though I did not get to contribute as much as I had hoped, I'm working on it) :smile: 

Regarding the maximum attention hook, it would be awesome if a gif was to contain multiple devices all lined up, all lights out, blinking. Unfortunately, I don't have that many pmos devices yet (I'm getting there, but slowly...)

> A graph for packaged devices over time

It might not meet the requirements here, but have you heard of [gource](http://gource.io/)? (I first heard of it trough this [blog post](https://laysrodriguesdev.wordpress.com/2018/04/20/almost-10-years-of-plasma-desktop/) on planet kde).

I have some experience with graphviz, but I don't really see the kind of representation you would like to obtain. Would that be a few tables side-by-side with the added devices in them? (Like, one table per month). Or more like a timeline? Extracting the device list itself seems quite easy to do :)

MartijnBraam commented:
I've written a python scripts that generates the data for a devices-over-time chart.

I think we actually want a chart, not a graph I think? @ollieparanoid

opendata26 commented:
@ollieparanoid no, on Sirius Sony have a modified modem so that it requires a ta service to 'post' so I'm waiting on Sony to allow Bjorn to release that, I have picture of Sirius on wiki page and more if you look in matrix history.

Personally I think you should mention some of the achievements made to do with mainline since that talk by Jeremy in the first paragraph, as it would make mainline sound even better, still keep modem bulk part in a different paragraph though.

One thing I think we should really highlight is that if they have decent knowledge of Linux and if they have a supported soc - I would say S4 Pro, Snapdragon 820 (you would want UART for that but once it's running most stuff should work), 845(same as 820, just need tons of patches on patchwork),  410, 801, 600 and probably S4 (non pro). That they can feel free to join the matrix and that it's very likely we'll end up with a working mainline kernel.

ollieparanoid commented:
Yeah, I had a chart like the one @MartijnBraam made in mind, x-axis is time and y-axis is the amount of devices. This is pretty good already! Only the time axis is kind of unreadable. Maybe display only the month there? If you don't mind it would be nice if you could upload the source as well.

> Personally I think you should mention some of the achievements made to do with mainline since that talk by Jeremy in the first paragraph

Which ones specifically? I've thought about listing more, but also I did not want to blow up the introduction.

Thanks for the quick feedback guys! \o/

opendata26 commented:
@ollieparanoid its SLiMbus, not slimport, its not merged yet and i would probably remove that whole thing to be honest as its kinda device specific due to amp and codec varies from soc to soc.

McBitter commented:
Short paragraph of what low-level has been up to.
We managed to dump brom of MT6735M and one more SoC using SWD (https://www.silabs.com/community/mcu/32-bit/knowledge-base.entry.html/2014/10/21/serial_wire_debugs-qKCT).
So the diagram for that is following: USBPHY -> set mux for accessing SWD (happens over USB D+ and D- pins, USB voltage level for signalling) -> Access SWD and speak CoreSight protocol on it. (https://www.arm.com/products/system-ip/coresight-debug-trace).
Some of the people have been writing documentation on what's known already.
Feel free to add more details.
@cyrozap @ollieparanoid

michitux commented:
I have created a (not so) short video of Gimp (under i3) on my Samsung Galaxy Note 8.0 tablet: https://youtu.be/4aRadGSQ_t4 There is certainly a lot that can be improved in this video (apart from the fact that somehow the pen is never recognized correctly when Gimp is started the first time). I could also demo other applications like evince for reading PDF files or (new) epiphany (see [here](https://github.com/yangxuan8282/rpi-aports/blob/master/epiphany/APKBUILD) for an APKBUILD) for web-browsing or xournal for taking notes and scribbling on PDF files. Unfortunately, not everything you can see in the video is nicely packaged for pmOS (the bar at the top needs https://github.com/postmarketOS/pmbootstrap/pull/1343 but I also do not want to suggest to merge it now as-is). If you want to add something like that (or another) demo video/GIF to the blog post I can also try to create a new, better video.

cyrozap commented:
@McBitter To clarify, I only dumped the MT6797 (Helio X20) BROM via SWD/JTAG, and I only used that method because it was convenient for me. If preloader sources for an SoC are available, like they were for the MT6737M-based Orange Pi 4G-IoT, it's a lot easier to dump the BROM from Linux after disabling the BROM/SRAM lockout code in the preloader--doing that doesn't require any special hardware and most boards don't have the default JTAG/SWD pins exposed anyways.

opendata26 commented:
Yeah,  Seems mostly fine,  only thing I would change is bamse is an ex Sony employee - now at linaro. And I would just say 820 instead of 820e

PureTryOut commented on content/blog/2018-05-26-one-year.md:
```diff
@@ -33,60 +33,74 @@ At this point, postmarketOS is only meant to be used by developers. In most inst
 
-
-Of course it would be nice if one could accept the call and actually talk to the other end. We're not quite there yet, but thanks to Srinivas Kandagatla who [submitted patches for the SLIMbus](https://lkml.org/lkml/2017/11/15/359) to mainline, the Linux kernel will soon be able to utilize the audio stack on Qualcomm phones like the Nexus 5. Then we can puzzle the userspace functionallity together ...
+Of course it would be nice if one could accept the call and actually talk to the other end. We're not quite there yet, but thanks to Srinivas Kandagatla who is working on mainline audio support for the [Snapdragon 820E SoC](https://en.wikipedia.org/wiki/List_of_Qualcomm_Snapdragon_devices#Snapdragon_820_and_821), the Linux kernel will soon be able to utilize the audio stack on Qualcomm phones lik ...
```
Last sentence, "functional**l**ity"

opendata26 commented:
I'm about 75% done with speakers working on Sirius with mainline, once o get it working it may be worth mentioning

cyrozap commented:
@ollieparanoid: To clarify a few things:
- We actually have three BROMs dumped now, not just one.
  - I've dumped the BROMs for both the MT6797 (Helio X20) and the MT6737M.
  - @McBitter dumped the BROM for the MT6735P.
- I actually just used JTAG (not SWD) to dump the BROM, but the specifics of the physical layer protocol (SWD vs. JTAG) really aren't that important--on this SoC and many others, both interfaces expose the same logical interface (the CoreSight peripheral) which can read from and write to device memory and memory-mapped peripherals.
- The reasons for dumping the BROMs are as follows (@McBitter, feel free to add to this list):
  - We want to understand the boot process with greater detail.
    - How is the boot metadata formatted on flash memory?
    - To what extent are any firmware signatures verified by the BROM? And how does it perform that verification?
    - Are there any vulnerabilities in the firmware verification process?
    - What functions/options do specific EFUSE/OTP locations and values control?
    - Other than the button to enter BROM download mode, are there any other methods for recovering from a bad flash?
  - We also want to understand what state the BROM leaves the chip in after it's finished executing.
    - Is any debugging functionality disabled by the BROM?
    - Are any write-once registers written to by the BROM? If so, what do they do?
    - Do any memory regions/registers have their access disabled by the BROM?
    - What are the preloader/future U-Boot SPL responsible for initializing?
  - Basically, since we don't have much documentation on how the BROM works (and since documentation often lies or isn't the whole truth, anyways), we need to discover for ourselves how it works in order to write our own documentation for it.

MartijnBraam commented:
New devices since the 219 days post:

```
samsung-kminilte
htc-m8
samsung-klte
zte-kis3
htc-a5ul
motorola-harpia
oneplus-oneplus2
sony-seagull
sony-sirius
xiaomi-santoni
infocus-flatfish
htc-flounder
xiaomi-mido
nextbit-robin
xiaomi-armani
sony-scorpion
samsung-i8190
motorola-peregrine
samsung-lt023g
semc-anzu
htc-vivo
samsung-i9195
motorola-falcon
motorola-shamu
lg-bullhead
samsung-gts210vewifi
samsung-manta
motorola-maserati
motorola-potter
samsung-kylessopen
jolla-sbj
asus-tf101
leeco-s2
motorola-athene
samsung-n5110
gp-peak
nokia-n9
asus-t00f
samsung-p4wifi
```

PureTryOut commented on content/blog/2018-05-26-one-year.md:
```diff
@@ -112,6 +112,48 @@ Speaking of free firmware: our recently established [#postmarketOS-lowlevel](/bl
 
 ## 39 New Devices (80 Total)
 
+<table>
```
Although I'm not sure if the markdown parser for the website supports it, at least the Github flavored Markdown supports tables itself. Do we really need plain HTML here?

MartijnBraam replied:
I just checked the previous post _after_ making this and now I understand why @ollieparanoid mentioned parsing the wiki. I'll have to update some code to make it link to the wiki pages.

ollieparanoid commented:
I've put all of my postmarketOS time for today into the blog post and got most of the images inserted (you wouldn't believe how much time this actually takes :p). Looks like we can get it done this out on this weekend (which means the writing must be finished on Thursday so craftyguy has time for the final review).

@MartijnBraam: thanks, that's super helpful!
@cyrozap: thanks for your clarifications, soon I'll update the post accordingly.

TODOs:
* Rewrite the lowlevel part with new information
* Add pictures to the new devices section
* Add the "Thanks to:" lines and make sure everyone is credited properly
* Maybe some more images towards the bottom sections which don't have images at all
* Gif from the maximum attention initramfs hook (@MartijnBraam?)

opendata26 commented:
maybe add some pictures of switch port?

zhuowei commented:
@opendata26 I only have that one photo right now; do you need more?

opendata26 commented:
thats probably enough, just to show it works

PureTryOut commented:
Maybe Plasma Mobile on it would be cool? Fail0verflow has been showing it with Plasma Desktop, so we can show it's mobile equivalent :smile:

ollieparanoid commented:
The switch port is mentioned in the outro text with a link to the PR with the photo:
> ### No End In Sight
> While these lines are being written, the next stunning pull requests are already rolling in (e.g. Nintendo Switch in [#1534](https://github.com/postmarketOS/pmbootstrap/pull/1534)). [...]

I would leave it at that TBH, but if you prefer we could also add the photo there.

MayeulC commented on content/blog/2018-05-26-one-year.md:
```diff
@@ -0,0 +1,258 @@
+* [Port your phone](https://wiki.postmarketos.org/wiki/Porting_to_a_new_device)
+* [Install pmOS](https://wiki.postmarketos.org/wiki/Installation_guide) on your machines if there's already a port
+* Make [good photos/videos](https://wiki.postmarketos.org/wiki/Making_good_photos) of devices running pmOS for the [wiki](https://wiki.postmarketos.org/wiki/Special:ListFiles) and [/r/postmarketOS](https://www.reddit.com/r/postmarketOS/)
+* [#1286](https://github.com/postmarketOS/pmbootstrap/issues/1286): packaging [Anbox](https://anbox.io/), and experimental project that runs Android apps on regular Linux distributions. After some patching we have it compiling and showing the "Starting..." screen (on the right), but the way it configures LXC does not work out of the box for us. To move forward, one could install Anbox on Ubuntu ( ...
```
I feel like (on the right) is not enough to point to the picture. (pictured/shown on the right)?

PureTryOut replied:
Do note that on mobile, the picture will not actually be on the right but either beneath or above. Not sure if we should take this into consideration in the text though.

PureTryOut commented on content/blog/2018-05-26-one-year.md:
```diff
@@ -0,0 +1,258 @@
+
+## Get Involved
+### Mainline Your Device!
+Mainlining expert [@opendata26](https://github.com/opendata26) encourages our readers to join the effort: "One thing I think we should really highlight is that if they have decent knowledge of Linux and if they have a supported SoC - I would say S4 Pro, Snapdragon 820 (you would want [UART](https://wiki.postmarketos.org/wiki/Serial_debugging) for that but once it's running most stuff should work) ...
```
Also, Matrix with a capital M

opendata26 replied:
He does, @andersso

opendata26 replied:
He does, @andersson

opendata26 commented:
I still think we should put switch picture on main post, as some people might not click it and it's pretty cool.

ollieparanoid commented:
@MartijnBraam and @MayeulC: thanks for adding the video! however, it looks like an anti-shaking filter was applied. so instead of the devices shaking, the frame of the video is shaking, which looks a bit weird. can this be corrected?

Thanks @daveloyall and @MayeulC, fixed everything you've found!
@opendata26: added the switch :+1: 
@craftyguy: the text is ready for the final edit! :grin: 

(if someone finds another issue with the text that is critical, we need to wait until the final edit is complete and fix it right before the release.)

daveloyall commented:
"HTC" is sometimes shown as "hTC".  Is that on purpose?

MartijnBraam commented:
@daveloyall It's in the source data somehow: https://github.com/postmarketOS/pmbootstrap/blob/master/aports/device/device-htc-a5ul/deviceinfo#L5

cyrozap commented on content/blog/2018-06-09-one-year.md:
```diff
@@ -0,0 +1,295 @@
+
+[![](/static/img/2018-06/pixelated-logo-thumb.jpg){: class="fr ml3 cr mb4"}](/static/img/2018-06/pixelated-logo.jpg)
+
+Speaking of free firmware: our recently established [#postmarketOS-lowlevel](/blog/2018/04/14/lowlevel/) crew is taking baby steps towards liberating bootloaders and cellular modem firmware for MediaTek devices. Around the time of the announcement blog post in April, quite a few people have joined the channel and started contributing. Notably [@cyrozap](https://github.com/cyrozap) has managed to  ...
```
The MT6737M BROM wasn't dumped via JTAG--only the MT6797 BROM was dumped that way. The MT6737M BROM was dumped by modifying the preloader so that it wouldn't lock access to the BROM, and then the BROM was dumped from `/dev/mem` in Linux.

I'd suggest changing

> ... of the MT6797 (Helio X20) and the MT6737M via [JTAG](https://en.wikipedia.org/wiki/JTAG).

to

> ... of the MT6797 (Helio X20) and the MT6737M. The first was dumped via [JTAG](https://en.wikipedia.org/wiki/JTAG), and the second was dumped from `/dev/mem` by modifying the preloader to remove some memory access restrictions that prevented Linux from reading the BROM.

and

> After proper reverse engineering, we should be able to tell how the metadata is formatted on the flash memory and possibly find new options to recover from a bad flash. A better understanding of the firmware signature verification by the BROM may allow us to find vulnerabilities and finally bypass these restrictions.

to

> After proper reverse engineering, we should better understand how boot metadata is formatted on the flash memory and possibly find new options to recover from a bad flash. In addition, a better understanding of the firmware signature verification process performed by the BROM could enable us to discover vulnerabilities and bypass these restrictions.

daveloyall replied:
Ok, see #63.

cyrozap commented on content/blog/2018-06-09-one-year.md:
```diff
@@ -0,0 +1,295 @@
+
+Speaking of free firmware: our recently established [#postmarketOS-lowlevel](/blog/2018/04/14/lowlevel/) crew is taking baby steps towards liberating bootloaders and cellular modem firmware for MediaTek devices. Around the time of the announcement blog post in April, quite a few people have joined the channel and started contributing. Notably [@cyrozap](https://github.com/cyrozap) has managed to  ...
+
+Towards replacing components in the low level boot process on these SoCs, we also want to understand what state the BROM leaves the chip in after it's finished executing. Whether the BROM disables debugging functionallity, makes use of write-once registers or disables memory regions and registers. That information could be used to use U-Boot as Secondary Program Loader (SPL) for initialization in ...
```
I'd suggest changing

> Whether the BROM disables debugging functionallity, makes use of write-once registers or disables memory regions and registers.

to

> For instance, we want to know if the BROM disables any debugging functionality, makes use of write-once registers, or disables memory regions and registers.

and change

> That information could be used to use U-Boot as Secondary Program Loader (SPL) for initialization in the future.

to

> That information might be useful for porting U-Boot SPL (Secondary Program Loader) to the platform, which would replace the proprietary preloader.

daveloyall replied:
ok, see #63.

craftyguy commented:
grammar/spelling stuff added. See my commit message for more details.

Also forgot to mention in the commit message, but I have applied the changes from here: https://github.com/postmarketOS/postmarketos.org/pull/63, so merging that PR is no longer necessary (@ollieparanoid requested I do this before editing)

daveloyall commented:
Ok @craftyguy, does that mean I can delete #63 and the associated branch in my account?

craftyguy commented:
@daveloyall yea, I applied that patch before editing, so it should be included in my commit

ollieparanoid commented:
Awesome work! I'm answering some messages, and then we should be able to release this in a few hours.
